src/Dictionaries/TrieDictionary.cpp

```diff
@@ -31,20 +24,175 @@ namespace ErrorCodes
+
+            return buf;
+        }
+
+        inline UInt8 prefixIPv6() const
+        {
+            return isv6 ? prefix : prefix + 96;
+        }
+    };
+
+    struct IPv4Subnet
+    {
+        UInt32 addr;
+        UInt8 prefix;
+    };
+
+    struct IPv6Subnet
+    {
+        const uint8_t * addr;
+        UInt8 prefix;
+    };
 }
 
 static void validateKeyTypes(const DataTypes & key_types)
 {
-    if (key_types.size() != 1)
-        throw Exception{"Expected a single IP address", ErrorCodes::TYPE_MISMATCH};
+    if (key_types.empty() || key_types.size() > 2)
+        throw Exception{"Expected a single IP address or IP with mask", ErrorCodes::TYPE_MISMATCH};
 
     const auto & actual_type = key_types[0]->getName();
-
     if (actual_type != "UInt32" && actual_type != "FixedString(16)")
         throw Exception{"Key does not match, expected either UInt32 or FixedString(16)", ErrorCodes::TYPE_MISMATCH};
+
+    if (key_types.size() > 1)
+    {
+        const auto * mask_col_type = typeid_cast<const DataTypeUInt8 *>(key_types[1].get());
+        if (mask_col_type == nullptr)
+            throw Exception{"Mask do not match, expected UInt8", ErrorCodes::TYPE_MISMATCH};
+    }
+}
+
+template <typename T, typename Comp>
+size_t sort_and_unique(std::vector<T> & vec, Comp comp)
+{
+    std::sort(vec.begin(), vec.end(),
+              [&](const auto & a, const auto & b) { return comp(a, b) < 0; });
+
+    auto new_end = std::unique(vec.begin(), vec.end(),
+                               [&](const auto & a, const auto & b) { return comp(a, b) == 0; });
+    if (new_end != vec.end())
+    {
+        vec.erase(new_end, vec.end());
+        return std::distance(new_end, vec.end());
+    }
+    return 0;
+}
+
+template <typename T>
+static inline int compareTo(T a, T b)
+{
+    return a > b ? 1 : (a < b ? -1 : 0);
+}
+
+inline static UInt32 IPv4AsUInt32(const void * addr)
+{
+    return Poco::ByteOrder::fromNetwork(*reinterpret_cast<const UInt32 *>(addr));
+}
+
+/// Convert mapped IPv6 to IPv4 if possible
+inline static UInt32 mappedIPv4ToBinary(const uint8_t * addr, bool & success)
+{
+    success = addr[0] == 0x0 && addr[1] == 0x0 &&
+              addr[2] == 0x0 && addr[3] == 0x0 &&
+              addr[4] == 0x0 && addr[5] == 0x0 &&
+              addr[6] == 0x0 && addr[7] == 0x0 &&
+              addr[8] == 0x0 && addr[9] == 0x0 &&
+              addr[10] == 0xff && addr[11] == 0xff;
+    if (!success)
+        return 0;
+    return IPv4AsUInt32(&addr[12]);
+}
+
+/// Convert IPv4 to IPv6-mapped and save results to buf
+inline static void mapIPv4ToIPv6(UInt32 addr, uint8_t * buf)
+{
+    memset(buf, 0, 10);
+    buf[10] = '\xFF';
+    buf[11] = '\xFF';
+    addr = Poco::ByteOrder::toNetwork(addr);
+    memcpy(&buf[12], reinterpret_cast<const uint8_t *>(&addr), 4);
+}
+
+static bool matchIPv4Subnet(UInt32 target, UInt32 addr, UInt8 prefix)
+{
+    UInt32 mask = (prefix >= 32) ? 0xffffffff : ~(0xffffffff >> prefix);
```

Comment:
Well, it is almost ok except that `0xffffffff` is an `int` literal, which could be 32 bit for some platforms.
Then,  `0xffffffff` could be signed, and, according to cppreference.com:
```
For signed lhs with nonnegative values, <...> , otherwise the behavior is undefined
```

So, it is better to use `0xffffffffu`